gocourse05/server.go

```diff
@@ -0,0 +1,39 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+
+	"gocourse05/camera"
+)
+
+const serverURL = "http://remote.animals.control"
+
+type Server struct {
+	Memory []camera.ProcessedData
+}
+
+func (s *Server) saveProcessedData(p camera.Processor) error {
+	processedData, err := p.ProcessData()
+	if err != nil {
+		return err
+	}
+
+	s.Memory = append(s.Memory, *processedData)
+
+	return nil
+}
```

Comment:
Пропоную перенести інтерфейс процесора в цей файл і подивитися, як працюватиме:

```suggestion
type Processor interface {
	// ...
}

type Server struct {
	Memory []camera.ProcessedData
	processor Processor
}

func NewServer(processor Processor) *Server {
	// ...
}

func (s *Server) saveProcessedData() error {
	processedData, err := s.processor.ProcessData()
	if err != nil {
		return err
	}

	s.Memory = append(s.Memory, *processedData)

	return nil
}
```

Пояснення тут https://go.dev/wiki/CodeReviewComments#interfaces

Reply:
Через те, що в різних пакетах камери перестають імплементувати методи, я не дуже розумію як правильно,
зробити все main і все з камери перенести в сервер?Воно якось все ломається з цією зміною.

І для кожної камери свій сервер виходить буде з NewServer то це якось не по завданню і не логічно виходить.

Reply:
[CodeReviewComments#interfaces](https://go.dev/wiki/CodeReviewComments#interfaces) каже:
> Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values.

Тобто інтерфейси оголошують там, де вони використовуються. В нашому випадку, інтерфейс `Processor` використовується сервером в файлі `server.go`.

Ось цей інтерфейс не дуже правильно оголошений, бо містить неекспортований метод `retrieveData`:
```go
type Processor interface {
	retrieveData() ([]Data, error)
	ProcessData() (*ProcessedData, error)
}

Я пропоную додати інтерфейс в файл `server.go`:

```
type Processor interface {
	ProcessData() (*ProcessedData, error)
}
	
type Server struct {
	// ...
	processor Processor
}
```

Всі камери, які мають метод `ProcessData() (*ProcessedData, error)`, автоматично імплементують цей інтерфейс.